File: src/main/java/in/costea/wiles/builders/ExpectParamsBuilder.java
package in.costea.wiles.builders;

import in.costea.wiles.enums.WhenRemoveToken;

import java.util.Collection;
import java.util.Objects;
import java.util.function.Predicate;

import static in.costea.wiles.statics.Constants.TOKENS_INVERSE;

public class ExpectParamsBuilder {


    public static final Predicate<String> ANYTHING = (x) -> true;
    private Predicate<String> foundTest;
    private String errorMessage = "Shouldn't happen";
    private WhenRemoveToken when = WhenRemoveToken.Default;
    private boolean ignoringNewLine = true;
    private ExpectParamsBuilder(Predicate<String> foundTest) {
        this.foundTest = foundTest;
    }

    public static Predicate<String> isContainedIn(Collection<String> set) {
        return set::contains;
    }

    public static ExpectParamsBuilder tokenOf(String expectedToken) {
        return new ExpectParamsBuilder(x -> Objects.equals(x, expectedToken))
                .withErrorMessage("Token \"" + TOKENS_INVERSE.get(expectedToken) + "\" expected!");
    }

    public static ExpectParamsBuilder tokenOf(Predicate<String> found) {
        return new ExpectParamsBuilder(found);
    }

    public ExpectParamsBuilder withErrorMessage(String message) {
        this.errorMessage = message;
        return this;
    }

    public ExpectParamsBuilder removeTokenWhen(WhenRemoveToken when) {
        this.when = when;
        return this;
    }

    public ExpectParamsBuilder dontIgnoreNewLine() {
        ignoringNewLine = false;
        return this;
    }

    public ExpectParamsBuilder or(Predicate<String> otherTest) {
        foundTest = foundTest.or(otherTest);
        return this;
    }

    @SuppressWarnings("unused")
    public ExpectParamsBuilder or(String token) {
        foundTest = foundTest.or(x -> x.equals(token));
        return this;
    }

    public Predicate<String> getFoundTest() {
        return foundTest;
    }

    public String getErrorMessage() {
        return errorMessage;
    }

    public WhenRemoveToken getWhenRemoveToken() {
        return when;
    }

    public boolean isIgnoringNewLine() {
        return ignoringNewLine;
    }
}

File: src/main/java/in/costea/wiles/builders/ExpectParamsBuilder.kt
package `in`.costea.wiles.builders

import `in`.costea.wiles.constants.ErrorMessages.FROZEN_ERROR
import `in`.costea.wiles.constants.ErrorMessages.ERROR_MESSAGE_EXPECTED_ERROR
import `in`.costea.wiles.enums.WhenRemoveToken
import `in`.costea.wiles.constants.Tokens.TOKENS_INVERSE
import `in`.costea.wiles.constants.ErrorMessages.TOKEN_EXPECTED_ERROR
import `in`.costea.wiles.constants.ErrorMessages.WHEN_REMOVE_EXPECTED_ERROR
import `in`.costea.wiles.exceptions.InternalErrorException
import java.util.function.Predicate

class ExpectParamsBuilder private constructor(var foundTest: Predicate<String>) {
    var errorMessage: String? = null
        private set
    var whenRemove: WhenRemoveToken = WhenRemoveToken.Default
        private set
    var isIgnoringNewLine = true
        private set
    var frozen = false
    private set
    fun withErrorMessage(message: String): ExpectParamsBuilder {
        checkFrozen()
        errorMessage = message
        return this
    }

    fun removeWhen(whenRemove: WhenRemoveToken): ExpectParamsBuilder {
        checkFrozen()
        this.whenRemove = whenRemove
        return this
    }

    fun dontIgnoreNewLine(): ExpectParamsBuilder {
        checkFrozen()
        isIgnoringNewLine = false
        return this
    }

    fun or(otherTest: Predicate<String>): ExpectParamsBuilder {
        checkFrozen()
        foundTest = foundTest.or(otherTest)
        errorMessage = null
        return this
    }

    fun or(otherTest: String): ExpectParamsBuilder {
        checkFrozen()
        or { x -> x == otherTest }
        errorMessage = null
        return this
    }

    fun freeze() : ExpectParamsBuilder
    {
        frozen = true
        if(errorMessage == null)
            throw InternalErrorException(ERROR_MESSAGE_EXPECTED_ERROR)
        if(whenRemove == WhenRemoveToken.Default)
            throw InternalErrorException(WHEN_REMOVE_EXPECTED_ERROR)
        return this
    }

    private fun checkFrozen()
    {
        if(frozen)
            throw InternalErrorException(FROZEN_ERROR)
    }

    companion object {
        @JvmStatic
        fun tokenOf(expectedToken: String): ExpectParamsBuilder {
            return ExpectParamsBuilder { x: String -> x == expectedToken }
                .withErrorMessage(TOKEN_EXPECTED_ERROR.format(TOKENS_INVERSE[expectedToken]))
        }

        @JvmStatic
        fun tokenOf(found: Predicate<String>): ExpectParamsBuilder {
            return ExpectParamsBuilder(found)
        }
    }
}